Inputs with type=color cannot be focused using keyboard tabs.

dgrammatiko opened the pull request:
#### ceaa50e0a560b9d37f5de1b1abd1f820ecbc15f5
<pre>
Inputs with type=color cannot be focused using keyboard tabs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266285">https://bugs.webkit.org/show_bug.cgi?id=266285</a>
<a href="https://rdar.apple.com/problem/119557550">rdar://problem/119557550</a>

Reviewed by NOBODY (OOPS!).

Removed the platform specific conditional for the ::isKeyboardFocusable (returned false for non-IOS)
Removed the `outline: none` from the html.css, to restore the visual indicator for the focus state

* LayoutTests/fast/forms/color/color-input-keyboard-focus-expected.txt: Added.
* LayoutTests/fast/forms/color/color-input-keyboard-focus.html: Added.
* Source/WebCore/css/html.css:
(#endif):
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::isKeyboardFocusable const):
(WebCore::ColorInputType::showPicker):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ceaa50e0a560b9d37f5de1b1abd1f820ecbc15f5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102166 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47610 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99130 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25158 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73955 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31169 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100088 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12826 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54296 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12579 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5665 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46940 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82644 "Found 10 new API test failures: TestWebKitAPI.ProcessSwap.NavigatingToLockdownMode, TestWebKitAPI.LockdownMode.NotAllowedFont, TestWebKitAPI.ServiceWorkers.LockdownModeInServiceWorkerProcess, TestWebKitAPI.LockdownMode.SVGFonts, TestWebKitAPI.LockdownMode.AllowedFont, TestWebKitAPI.ProcessSwap.LockdownModeSystemSettingChange, TestWebKitAPI.LockdownMode.NotSupportedFontLoadingAPI, TestWebKitAPI.LockdownMode.AllowedFontLoadingAPI, TestWebKitAPI.ProcessSwap.CannotDisableLockdownModeWithoutBrowserEntitlement, TestWebKitAPI.LockdownMode.NotAllowedFontLoadingAPI (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5742 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104188 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24159 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17622 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83006 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24538 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83916 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82407 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27080 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4638 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17664 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24124 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29275 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23947 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27259 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25520 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ad88ef728e9294a77522401f27a094e1f0784db7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

lukewarlow commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -269,7 +265,7 @@ void ColorInputType::handleDOMActivateEvent(Event& event)
 }
 
-void ColorInputType::showPicker() 
+void ColorInputType::showPicker()
```
What changed here? It looks identical. Could you try and drop whatever the change was?

dgrammatiko replied:
There was an extra space after the closing parenthesis. I can revert it but I guess it was the linter (?)

lukewarlow replied:
Ah okay (GitHub diff doesn't show that very well) I guess leave it changed else it'll just end up being changed by the next change to the file.

lukewarlow reviewed (commented):
Could you update the commit message with a simple explanation of this change.

That the iOS special casing is removed and the outline style was removed?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d1e321f0e56693c35b21f3728c0e28c509761799. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/de1a5b760ed59f2a9b36fca6c895dbe0ae4e181a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

lukewarlow commented:
You need to remove the `Explanation of why this fixes the bug (OOPS!).` line too. Also "rrestore" -> "restore" :)

lukewarlow commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -156,11 +156,7 @@ bool ColorInputType::isMouseFocusable() const
 {
     ASSERT(element());
-#if PLATFORM(IOS_FAMILY)
     return element()->isTextFormControlFocusable();
```
Pre-existing but because you touch it last you get to fix it.
```suggestion
    return protectedElement()->isTextFormControlFocusable();
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c0c309621a438276a71f2f977e04458c6191dc29. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8b9a36bc0187005d6a3ee3aa0558aaf7da7c1918. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

lukewarlow commented:
I've just noticed the top block of the commit message is duplicated, if you could fix that then I think this is ready for review.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/001511d07520d10e7bdf0a242981ea7df7d6c90d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8eac4d4e48a9b5a19515f74e393f9c571a1178c9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e7b72b9fe30c8bcf8af8e872a9d29f4d847d346f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8f65e62b6a271333ea436b8cf1a183d41dcc263c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1d19f2701514090560ab009f8e511bef8caa429f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

dgrammatiko committed:
Inputs with type=color cannot be focused using keyboard tabs.
https://bugs.webkit.org/show_bug.cgi?id=266285
rdar://problem/119557550

Reviewed by NOBODY (OOPS!).

Removed the platform specific conditional for the ::isKeyboardFocusable (returned false for non-IOS)
Removed the `outline: none` from the html.css, to restore the visual indicator for the focus state

* LayoutTests/fast/forms/color/color-input-keyboard-focus-expected.txt: Added.
* LayoutTests/fast/forms/color/color-input-keyboard-focus.html: Added.
* Source/WebCore/css/html.css:
(#endif):
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::ColorInputType::isKeyboardFocusable const):
(WebCore::ColorInputType::showPicker):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ceaa50e0a560b9d37f5de1b1abd1f820ecbc15f5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42910

annevk commented:
Are we sure this matches platform conventions for this control?

pxlcoder commented:
> Are we sure this matches platform conventions for this control?

I don't think it does. Color inputs are effectively buttons on macOS, which are not keyboard focusable by default.

These controls are only keyboard focusable if certain accessibility settings are turned on. Either in Safari settings or System settings.

lukewarlow commented:
Should this instead use the isTextFormControlKeyboardFocusable(event) function?

dgrammatiko commented:
@annevk @pxlcoder I guess the button is part of the HTML of a page and should follow the HTML specs. The part that is OS driven is the picker which is not(?) really affected here

How about

```
    if (!InputType::isKeyboardFocusable(event))
        return false;

    return protectedElement()->isTextFormControlFocusable();
```

lukewarlow commented:
What does the button type method do in this case?

dgrammatiko commented:
@lukewarlow https://github.com/WebKit/WebKit/blob/3ce45e50a24ce2949889ffa7d9f0f3cdb4f654a8/Source/WebCore/html/InputType.cpp#L674-L679

Seems reasonable but I guess the point about the platform is correct, buttons and input (type= file, submit, etc) are not focusable on MacOS without user opt in

lukewarlow commented:
Yeah so I think returning protectedElement()->isTextFormControlKeyboardFocusable(event) instead of return false is the correct approach. Leaving iOS as it is currently.

Buttons don't seem to change that from the base InputType impl.
